nav2_amcl/src/amcl_node.cpp

```diff
@@ -914,291 +742,262 @@ AmclNode::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
-  dynamic_param_client_->get_event_param("beam_skip_distance", beam_skip_distance_);
-  dynamic_param_client_->get_event_param("beam_skip_threshold", beam_skip_threshold_);
+  map->cells =
+    reinterpret_cast<map_cell_t *>(malloc(sizeof(map_cell_t) * map->size_x * map->size_y));
 
-  if (pf_ != NULL) {
-    pf_free(pf_);
-    pf_ = NULL;
+  // Convert to player format
+  for (int i = 0; i < map->size_x * map->size_y; i++) {
+    if (map_msg.data[i] == 0) {
+      map->cells[i].occ_state = -1;
+    } else if (map_msg.data[i] == 100) {
+      map->cells[i].occ_state = +1;
+    } else {
+      map->cells[i].occ_state = 0;
+    }
   }
+
+  return map;
+}
+
+void
+AmclNode::initTransforms()
+{
+  RCLCPP_INFO(get_logger(), "initTransforms");
+
+  // Initialize transform listener and broadcaster
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(rclcpp_node_->get_clock());
+  tf_buffer_->setUsingDedicatedThread(true);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, rclcpp_node_, false);
+  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(rclcpp_node_);
+
+  sent_first_transform_ = false;
+  latest_tf_valid_ = false;
+  memset(&latest_tf_, 0, sizeof(latest_tf_));
+}
+
+void
+AmclNode::initMessageFilters()
+{
+  laser_scan_sub_ = std::make_unique<message_filters::Subscriber<sensor_msgs::msg::LaserScan>>(
+    rclcpp_node_.get(), scan_topic_, rmw_qos_profile_sensor_data);
+
+  laser_scan_filter_ = std::make_unique<tf2_ros::MessageFilter<sensor_msgs::msg::LaserScan>>(
+    *laser_scan_sub_, *tf_buffer_, odom_frame_id_, 10, rclcpp_node_);
+
+  laser_scan_connection_ = laser_scan_filter_->registerCallback(std::bind(&AmclNode::laserReceived,
+      this, std::placeholders::_1));
+}
+
+void
+AmclNode::initPubSub()
+{
+  RCLCPP_INFO(get_logger(), "initPubSub");
+
+  particlecloud_pub_ = create_publisher<geometry_msgs::msg::PoseArray>("particlecloud",
+      rmw_qos_profile_sensor_data);
+
+  rmw_qos_profile_t amcl_qos_profile = rmw_qos_profile_default;
+  amcl_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+
+  pose_pub_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("amcl_pose",
+      amcl_qos_profile);
+
+  initial_pose_sub_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "initialpose", std::bind(&AmclNode::initialPoseReceived, this, std::placeholders::_1));
+}
+
+void
+AmclNode::initServices()
+{
+  global_loc_srv_ = create_service<std_srvs::srv::Empty>("global_localization",
+      std::bind(&AmclNode::globalLocalizationCallback, this, _1, _2, _3));
+
+  nomotion_update_srv_ = create_service<std_srvs::srv::Empty>("request_nomotion_update",
+      std::bind(&AmclNode::nomotionUpdateCallback, this, _1, _2, _3));
+}
+
+void
+AmclNode::initOdometry()
+{
+  // When pausing and resuming, remember the last robot pose so we don't start at 0:0 again
+  init_pose_[0] = last_published_pose_.pose.pose.position.x;
```

Comment:
At startup, `last_published_pose_` is not initialized.

Reply:
@mhpanah  When it first starts up, the last_published_pose_ is 0,0,0. One it runs, the last_published_pose is updasted. If AMCL goes through the deactivate and cleanup transitions and is restarted with configure and activate, it will remember the last_published pose. It seems this is the best it can do without a solution for external persistance. What do you think?

Reply:
@mjeronimo Yes, that's a good solution without external persistence.  But I don't see anywhere that `last_published_pose_` is set to 0,0,0 for start up. In the header file we are creating this variable without initializing it to 0,0,0.

Reply:
Ah, OK. The ROS IDL code generation automatically handles initializing fields for messages. In this case, it generates code like this:

```C++
namespace geometry_msgs
{

namespace msg
{

// message struct
template<class ContainerAllocator>
struct Point_
{
  using Type = Point_<ContainerAllocator>;

  explicit Point_(rosidl_generator_cpp::MessageInitialization _init = rosidl_generator_cpp::MessageInitialization::ALL)
  {
    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
    {
      this->x = 0.0;
      this->y = 0.0;
      this->z = 0.0;
    }
  }

  explicit Point_(const ContainerAllocator & _alloc, rosidl_generator_cpp::MessageInitialization _init = rosidl_generator_cpp::MessageInitialization::ALL)
  {
    (void)_alloc;
    if (rosidl_generator_cpp::MessageInitialization::ALL == _init ||
      rosidl_generator_cpp::MessageInitialization::ZERO == _init)
    {
      this->x = 0.0;
      this->y = 0.0;
      this->z = 0.0;
    }
  }
```

Reply:
Got it. Thanks.